File: 137. Single Number II/Solution.java
// LeetCode #137: Single Number II

/*
Given an array of integers, every element appears three times except for one. Find that single one.

Note:
Your algorithm should have a linear runtime complexity. Could you implement it without using extra memory?
*/

public class Solution {
    public int singleNumber(int[] nums) {
        int res = 0;
        for (int i = 0; i < 32; i++) {
            int bitSum = 0;
            for (int j = 0; j < nums.length; j++) {
                bitSum += (nums[j] >> i) & 1;
            }
            
            res = res | ((bitSum % 3) << i);
        }
        return res;
    }
}

File: 137. Single Number II/Solution.py
# LeetCode #137: Single Number II

'''
Given an array of integers, every element appears three times except for one. Find that single one.

Note:
Your algorithm should have a linear runtime complexity. Could you implement it without using extra memory?
'''

class Solution:
    # @param A, a list of integer
    # @return an integer
    def singleNumber(self, A):
        digits = [0] * 32
        for i in range(0, 32):
            for j in range(0, len(A)):
                digits[i] += (A[j] >> i) & 1
            
        res = 0    
        for k in range(0, 32):
            res += (digits[k] % 3) << k
            
        return res